### ISSUE TITLE
Forget history position

### ISSUE THREAD

**Issue by nu8** [malxau/yori#38]
Currently I using Bash with Cygwin. With Bash, I can enter these commands:

~~~
$ date
Thu, Apr 23, 2020  7:51:43 PM

$ echo Sunday
Sunday
~~~

Then if I press Up twice, I get this, as expected:

~~~
$ date
~~~

If I want to cancel, I can press down twice, or Ctrl + C. After I have canceled,
if I press Up again, it gives me the most recent result, as expected:

~~~
$ echo Sunday
~~~

Now contrast with Yori. I can enter these commands:

~~~
C:\yori-core-amd64»ydate
20200423
C:\yori-core-amd64»echo Sunday
Sunday
~~~

Then if I press Up twice, I get this, as expected:

~~~
C:\yori-core-amd64»ydate
~~~

but if I want to cancel, I cant press down twice. It only lets you press down
once, so that the most recent command displays:

~~~
C:\yori-core-amd64»echo Sunday
~~~

as a workaround, you can press Esc. But then if you press Up, you dont get the
most recent result anymore. Because it didnt let you scroll all the way down,
you get this:

~~~
C:\yori-core-amd64»ydate
~~~

As a workaround, instead of pressing Up, you can Ctrl + C like Bash for expected
results. I dont want the shell trying to be smart, and remembering the position
in the history. Does Yori have some option to change this behavior?

**Comment by malxau** [comment]
I pushed commit b26d2d2aefb51e525ed1efef96bae45368fbca2d for this.

For what it's worth, the last thing I was trying to be is "smart" here.  CMD has logic where if you press up twice and enter to execute a command, the position in history is remembered, so there are commands in history if you press down on the next command.  I always found this not very intuitive and this was one area I was trying to be stupider.  I don't mind going further if that's what people want...

**Comment by muv1** [comment]
Ok thanks. It seems no CI for this project, so I will have to wait or try to build it myself. However I dont have a Visual Studio environment so it might be difficult.

**Comment by malxau** [comment]
Do you have experience with CI or know what would be involved in setting it up? I looked at one once, and obviously by nature they need to be a little restrictive about the build tools they support; the one I looked at really wanted an msbuild project which isn't ideal since the compilers I normally use are much older, so I need to keep Makefile-based builds around anyway.

In terms of building it yourself, I use the [Visual C++ Build Tools](https://visualstudio.microsoft.com/visual-cpp-build-tools/) which are a free download. Yori should compile with any Windows SDK/compiler version.  The main issue is that the SDK is a pretty large download for a pretty small project.  Nonetheless, building Yori should just involve cd-ing to its source directory and running NMAKE from the command line.

**Comment by muv1** [comment]
Yeah, I have been looking for some time at creating my own small distribution of the Build Tools:

https://github.com/muv1/vase

as I hate the large site of even just the Build Tools. Until then I prefer projects that use GCC or Clang. This is my compiler of choice:

https://github.com/mstorsjo/llvm-mingw

**Comment by malxau** [comment]
It would be fantastic to get a specified, small download of build tools.  A project like this uses a tiny portion of it.  Reading the links was a bit sad too - it's not just that .NET is downloaded redundantly to support msbuild, but that Yori didn't need msbuild to begin with.  The normal download UI ends up with four versions of the compiler, for x86/x64 host with x86/x64 target, even though a normal developer only wants one or two.  I'd need to double check which import libraries Yori uses, but since it has its own mini-CRT and dynamically loads most functions, I think all it needs is a compiler/linker/nmake, kernel32.lib and possibly a handful of others, and a regrettably large set of includes to support windows.h.  I'll bet I could hand-craft an install that's under 100Mb.

The other thing to note about all of these tools is they compress incredibly well.  Includes and import libraries frequently are reduced in size by 75% or more.  That's not just a win for distribution, but there's no good reason to store these in uncompressed form after install either.  I end up using either per-file wof compression (see ```ycompact.exe /?```) for objects that are updated frequently, or even WIM backed compression for tools that will never be updated, which can yield enormous disk space savings.

Personally I think clang is very nice these days too.  The main reason for using Visual C++ is because I'm using older versions of it to target older versions of Windows.  In the past I've added support for gcc/clang later in projects, but it always ends up with multiple build systems, which is kind of unfortunate.  I know there are things like cmake which can rationalize some of this, but that just adds to the number of tools people need to work with the source.

**Comment by muv1** [comment]
A small version of Build Tools is possible. Its just a matter of figuring out what files are needed, then what packages have those files. Then you extract everything and put into a new Zip.

I actually did just that years ago, it was under 200 MB as you said. I was able to compile the Notepad2 program with it. But at the time it required all these Environment variables and Registry entries. It just got too complicated for me. But now I know a lot more, so I will probably take another crack at it.

**Comment by malxau** [comment]
I've been thinking about this all afternoon, and wondering how small a Yori build environment can be.  I started with the Windows Server 2003 SP1 SDK, which is the first official x64 compiler, halfway between 2003 and 2005.  I shoehorned in the Visual C++ 2003 x86 compiler from the Visual C++ Toolkit.  Both of these are (were?) freely downloadable.

Compressing the tree into a WIM results in a 5.25Mb WIM for both x86 and x64.  With `dism /apply-image /wimboot` , so files aren't extracted, that's basically the disk footprint too.

It shouldn't need any registry keys to run; it will need `PATH`, `INCLUDE`, and `LIB` set appropriately.  Ideally there should be a hacked up setenv.cmd in here for that.

```
Y:\>dir /s

 Directory of Y:\bin\x64\1033

04/25/2020  08:03 PM  <DIR>             .
04/25/2020  08:03 PM  <DIR>             ..
04/14/2005  07:11 PM            372,736 clui.dll
04/14/2005  07:11 PM             50,688 linkui.dll
           2 File(s)            423,424 bytes
           2 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\bin\x64

04/25/2020  08:03 PM  <DIR>             .
04/25/2020  08:03 PM  <DIR>             ..
04/25/2020  08:03 PM  <DIR>             1033
04/14/2005  07:11 PM            544,768 c1.dll
04/14/2005  07:11 PM          2,043,904 c2.dll
04/14/2005  07:11 PM             90,112 cl.exe
04/04/2005  07:43 PM             17,920 cvtres.exe
04/14/2005  07:11 PM              5,120 lib.exe
04/14/2005  07:11 PM            507,904 link.exe
04/14/2005  07:11 PM             53,248 msobj80.dll
04/14/2005  07:11 PM            247,808 mspdb80.dll
04/14/2005  07:11 PM            520,192 msvcp80.dll
04/14/2005  07:11 PM            491,520 msvcr80.dll
          10 File(s)          4,522,496 bytes
           3 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\bin\x86

04/25/2020  08:10 PM  <DIR>             .
04/25/2020  08:10 PM  <DIR>             ..
03/18/2003  09:11 PM            933,888 c1.dll
03/18/2003  09:12 PM          1,867,776 c2.dll
03/18/2003  08:37 PM             86,016 cl.exe
03/15/2003  01:46 AM                145 cl.exe.config
04/04/2005  07:43 PM             17,920 cvtres.exe
03/18/2003  08:38 PM            647,168 link.exe
03/15/2003  01:46 AM                145 link.exe.config
03/18/2003  08:12 PM             73,728 msobj71.dll
03/18/2003  08:09 PM            241,664 mspdb71.dll
03/18/2003  09:14 PM            499,712 msvcp71.dll
04/14/2005  07:11 PM            520,192 msvcp80.dll
02/21/2003  05:42 AM            348,160 msvcr71.dll
04/14/2005  07:11 PM            491,520 msvcr80.dll
          13 File(s)          5,728,034 bytes
           2 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\bin

04/25/2020  07:37 PM  <DIR>             .
04/25/2020  07:37 PM  <DIR>             ..
04/14/2005  07:12 PM             59,392 nmake.exe
04/14/2005  06:59 PM              5,120 RC.Exe
04/14/2005  06:59 PM            135,680 RcDll.Dll
04/25/2020  08:03 PM  <DIR>             x64
04/25/2020  08:10 PM  <DIR>             x86
           3 File(s)            200,192 bytes
           4 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\inc

04/25/2020  07:41 PM  <DIR>             .
04/25/2020  07:41 PM  <DIR>             ..
04/14/2005  06:54 PM              8,998 BaseTsd.h
04/14/2005  06:54 PM              8,401 CommCtrl.rh
04/14/2005  06:54 PM             10,312 ctype.h
04/14/2005  06:54 PM                801 Dde.rh
04/14/2005  06:54 PM              6,136 Dlgs.h
04/14/2005  06:54 PM              3,293 excpt.h
04/14/2005  06:54 PM              5,079 Guiddef.h
04/14/2005  06:54 PM             26,511 Imm.h
04/14/2005  06:54 PM             24,465 Mcx.h
04/14/2005  06:54 PM              1,018 PopPack.h
04/14/2005  06:54 PM                915 PshPack1.h
04/14/2005  06:54 PM                915 PshPack2.h
04/14/2005  06:54 PM                915 PshPack4.h
04/14/2005  06:54 PM                915 PshPack8.h
04/14/2005  06:54 PM              4,762 Reason.h
04/14/2005  06:54 PM             39,122 specstrings.h
04/14/2005  06:54 PM              3,825 stdarg.h
04/14/2005  06:54 PM             14,778 StrAlign.h
04/14/2005  06:54 PM              7,803 string.h
04/14/2005  06:55 PM              3,917 Tvout.h
04/14/2005  06:55 PM            246,998 WinBase.h
04/14/2005  06:55 PM             20,529 WinCon.h
04/14/2005  06:55 PM              9,399 WinDef.h
04/14/2005  06:55 PM              6,334 Windows.h
04/14/2005  06:55 PM            616,740 WinError.h
04/14/2005  06:55 PM            200,805 WinGDI.h
04/14/2005  06:55 PM            126,129 WinIoCtl.h
04/14/2005  06:55 PM             26,251 WinNetWk.h
04/14/2005  06:55 PM             60,362 WinNls.h
04/14/2005  06:55 PM            321,384 WinNT.h
04/14/2005  06:55 PM             12,765 WinNT.rh
04/14/2005  06:55 PM             23,343 WinReg.h
04/14/2005  06:55 PM                622 WinResrc.h
04/14/2005  06:55 PM             31,190 WinSvc.h
04/14/2005  06:55 PM            294,296 WinUser.h
04/14/2005  06:55 PM             49,735 WinUser.rh
04/14/2005  06:55 PM              9,806 WinVer.h
          37 File(s)          2,229,569 bytes
           2 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\lib\x64

04/25/2020  07:43 PM  <DIR>             .
04/25/2020  07:43 PM  <DIR>             ..
04/14/2005  06:55 PM            141,938 AdvAPI32.Lib
04/14/2005  06:55 PM              6,784 ComDlg32.Lib
04/14/2005  06:55 PM             74,308 Gdi32.Lib
04/14/2005  06:55 PM            184,196 Kernel32.Lib
04/14/2005  06:55 PM            132,540 User32.Lib
           5 File(s)            539,766 bytes
           2 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\lib\x86

04/25/2020  07:50 PM  <DIR>             .
04/25/2020  07:50 PM  <DIR>             ..
04/14/2005  06:55 PM            153,384 AdvAPI32.Lib
04/14/2005  06:55 PM              7,220 ComDlg32.Lib
04/14/2005  06:55 PM             80,910 Gdi32.Lib
04/14/2005  06:55 PM            196,054 Kernel32.Lib
04/14/2005  06:56 PM            142,150 User32.Lib
           5 File(s)            579,718 bytes
           2 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\lib

04/25/2020  07:26 PM  <DIR>             .
04/25/2020  07:26 PM  <DIR>             ..
04/25/2020  07:43 PM  <DIR>             x64
04/25/2020  07:50 PM  <DIR>             x86
           0 File(s)                  0 bytes
           4 Dir(s)      70,466,080,768 bytes free

 Directory of Y:\

04/25/2020  07:20 PM  <DIR>             .
04/25/2020  07:20 PM  <DIR>             ..
04/25/2020  07:37 PM  <DIR>             bin
04/25/2020  07:41 PM  <DIR>             inc
04/25/2020  07:26 PM  <DIR>             lib
           0 File(s)                  0 bytes
           5 Dir(s)      70,466,080,768 bytes free

     Total Files Listed:
          75 File(s)         14,223,199 bytes
          26 Dir(s)

...

 Directory of Y:\

04/25/2020  08:11 PM          5,505,700 yoribuild.wim
           1 File(s)          5,505,700 bytes
           0 Dir(s)      70,466,080,768 bytes free

```

**Comment by muv1** [comment]
~~~
04/25/2020  08:11 PM          5,505,700 yoribuild.wim
~~~

...Jesus Christ. I would be interested to try that out. My concern is that first, its the 2003 version, so after some sanity checks probably I would still want to try with newer version. Second, I would be concerned it wont even compile larger programs. For example I want to compile a static C LibCurl to be used with a D program.

At any rate, very cool! You should put the WIM here or somewhere:

https://github.com/malxau/yori/releases

**Comment by malxau** [comment]
Obviously this package is more of a proof of concept than anything practical.  It's removing everything Yori doesn't need, so any other software would need things added back.  Note specifically it doesn't have any C runtime support; the import libraries for the C runtime DLLs are fairly small since they don't contain code, but the static versions are large and get larger with each compiler revision.  Looking at the 18362 SDK, libucrt.lib is 63Mb for x64 and 61Mb for x86, and those are only usable with an extra 1Mb of headers. To put that in perspective, libcmt.lib from the 2003 SP1 SDK is 5Mb.  Compression gets a lot of that back, but that's the difference between a 200Mb package using current tools for general C programs and a 5Mb package for Yori specifically.

**Comment by muv1** [comment]
Would you be interested in making a "small build tools" package for the current version of Visual Studio? I plan on doing myself but I am busy with other things. I so, I would be willing to donate $1,000 for your time. I can put some up front. Let me know if you are interested svnpenn@gmail.com

**Comment by malxau** [comment]
Per the [license agreement](https://visualstudio.microsoft.com/license-terms/mlt031519/) in section 4, I'm not allowed to provide the software independently - this is why I detailed what I did and how I did it, so that others are free to replicate it, without publishing a package of the result.

I'm happy to track down the people in Visual Studio and pitch to them different ways of packaging and distributing build tools that would be more useful for smaller software projects.  Just note though where they are coming from - the amount of tools they're offering is growing, as is the amount of storage available to store them, and the size of networks to transport them.  Trying to split things up into packages of a few Mb would result in many thousands of components that would be very difficult to test together and very hard for users to know what they need.  Note that older versions offer a similar number of options to current versions, but each option ends up being far more detailed because there is less code in total - so Visual C++ 6 et al allow you to independently select each of six available C runtime libraries, for instance.

So, what would be the "ideal" way that build tools would be composed, in your opinion?

**Comment by muv1** [comment]
I just want something that include the C compiler (I beleve `cl.exe`) as well as the make tool (I believe `nmake.exe`), as well as the standard headers and libraries. I only need the command line tools, just enough so I could compile a C file, or handle a project that provides an Nmake file.

**Comment by malxau** [comment]
I followed up with the Visual Studio team.  They're hearing similar feedback from Build Tools users who want to compile in limited environments such as containers, although don't have any concrete announcements to make at this time.  Nonetheless the feedback has been heard.

**Comment by muv1** [comment]
You wouldnt have to distribute a binary package. If you just released a powershell script or something that creates the small package, that might be an acceptable solution

**Comment by malxau** [comment]
The original issue here has been fixed in the 1.30 release.  Please reactivate this issue or file a new issue if there are further issues related to it.
